botocore/credentials.py

```diff
@@ -791,6 +820,193 @@ def _assume_role_base_kwargs(self, config):
+            **self._assume_role_base_kwargs(self._get_role_config_values()))
+        creds = self._create_creds_from_response(response)
+        return creds, response
+
+    def _assume_role_base_kwargs(self, config):
+        assertion = None
+        for authenticator in self.authenticators:
+            if authenticator.is_suitable(config):
+                assertion = authenticator.authenticate(
+                    config=config, username_prompter=self.username_prompter,
+                    password_prompter=self.password_prompter)
+                break
+        else:
+            raise ValueError("Unsupported saml_authentication_type: %s"
+                             % config.get('saml_authentication_type'))
+        if not assertion:
+            raise SamlError(detail='Login failed: SAML assertion not found')
+        idp_roles = self._parse_roles(assertion)
+        if not idp_roles:
+            raise SamlError(detail='Identity provider provides no role.')
+        role = self.role_selector(config.get('role_arn'), idp_roles)
+        if not role:
+            raise SamlError(detail='Unable to choose role "%s" from %s' % (
+                config.get('role_arn'), [r['RoleArn'] for r in idp_roles]))
+        role['SAMLAssertion'] = assertion
+        return role
+
+    def _parse_roles(self, assertion):
+        attribute = '{urn:oasis:names:tc:SAML:2.0:assertion}Attribute'
+        attr_value = '{urn:oasis:names:tc:SAML:2.0:assertion}AttributeValue'
+        awsroles = []
+        root = ET.fromstring(base64.b64decode(assertion))
+        for attr in root.getiterator(attribute):
+            if attr.get('Name')=='https://aws.amazon.com/SAML/Attributes/Role':
+                for value in attr.getiterator(attr_value):
+                    parts = value.text.split(',')
+                    # Deals with "role_arn,pricipal_arn" or its reversed order
+                    if 'saml-provider' in parts[0]:
+                        role = {'PrincipalArn': parts[0], 'RoleArn': parts[1]}
+                    else:
+                        role = {'PrincipalArn': parts[1], 'RoleArn': parts[0]}
+                    awsroles.append(role)
+        return awsroles
+
+
+class SamlAuthenticator(object):
+    def is_suitable(self, config):
+        """Return True if this instance intends to perform authentication"""
+        raise NotImplemented()
+
+    def authenticate(self, config, username_prompter, password_prompter):
+        """Returns SAML assertion when login succeeds, or None otherwise"""
+        raise NotImplemented()
+
+
+class SamlGenericFormsBasedAuthenticator(SamlAuthenticator):
+    username_field = 'username'
+    password_field = 'password'
+
+    def is_suitable(self, config):
+        return config.get('saml_authentication_type') == 'form'
+
+    def authenticate(self, config, username_prompter, password_prompter):
+        verify = config.get('saml_verify_ssl') != 'false'
```

Comment:
I don't have preference here. Just want to share the background here, so that we make an informed decision.

This (undocumented) option is necessary, only when authenticating against some testing SAML IdP setup. This would probably happen again, when customer administrator is setting up their internal SAML IdP from scratch, without a SSL certificate yet, and wanting to test its function really quick. Besides, it will be the CLI customer who setup that most important "saml_endpoint" setting in their `~/.aws/credentials`, so it seems logically true that they should know what IdP they are sending their login credentials to. If they don't trust an IdP, they don't use it in the first place. An "saml_verify_ssl = false" just provides options when needed.

Let me know your preference. I am all ears. Meanwhile I'm finding some security review too.